index.js

```diff
@@ -1,86 +1,53 @@
 const async = require('async');
 
-module.exports = (collection, schema, method, callback) => {
+module.exports = async(collection, schema, method) => {
   const key = schema.key;
   const property = schema.set;
   const get = schema.get;
-  const cache = {};
 
-  // assigns either a specific field or the whole object:
-  // for either an array or a field:
-  const assign = (destination, source) => {
-    const element = get ? source[get] : source;
-    if (Array.isArray(destination)) {
-      destination.push(element);
-    } else {
-      destination[property] = element;
-    }
-  };
-
-  // applies the fetch method one time to one item:
-  const applySingleMethod = (collectionItem, done) => {
-    if (cache[collectionItem[key]]) {
-      assign(collectionItem, cache[collectionItem[key]]);
-      return done();
-    }
-    return method(collectionItem[key], (err, fetchedItem) => {
-      if (err) {
-        return done(err);
-      }
-      cache[collectionItem[key]] = fetchedItem;
-      assign(collectionItem, fetchedItem);
-      done();
-    });
-  };
+  /// make sure collection is an array:
+  if (!Array.isArray(collection)) {
+    collection = [collection];
+  }
 
-  // applies the fetch method one time each to an array of items:
-  const applyMultipleMethod = (collectionItem, done) => {
-    const allFetchedItems = [];
-    async.eachSeries(collectionItem[key], (listItem, eachDone) => {
-      if (cache[listItem]) {
-        assign(allFetchedItems, cache[listItem]);
-        return eachDone();
-      }
-      method(listItem, (err, fetchedItem) => {
-        if (err) {
-          return done(err);
+  // get unique list of keys in the collection:
+  const uniqueKeys = collection.reduce((accumulator, item) => {
```

Comment:
We actually could reduce to an obj of key and promise and then run through p-prop